lib/vintage_net_lte/at_command.ex

```diff
@@ -0,0 +1,136 @@
+defmodule VintageNetLTE.ATCommand do
+  @moduledoc """
+  Send AT commands to modem
+  """
+
+  @typedoc """
+  The options for the ATCommand server are:
+
+  * `:speed` - the speed of the serial connection
+  * `:tty` - the tty to send AT commands to
+  """
+  @type opt :: {:speed, non_neg_integer()} | {:tty, String.t()}
+
+  use GenServer
+  alias Circuits.UART
+
+  defmodule State do
+    @moduledoc false
+
+    defstruct uart: nil,
+              tty: nil,
+              buffer: [],
+              waiter: nil,
+              stop_response: nil,
+              command_timeout_ref: nil,
+              queue: :queue.new()
+  end
+
+  @spec start_link([opt]) :: GenServer.on_start()
+  def start_link(opts) do
+    tty = Keyword.fetch!(opts, :tty)
+    GenServer.start_link(__MODULE__, opts, name: server_name(tty))
+  end
+
+  @doc """
+  Send an AT command to the modem
+
+  This takes the tty name, the AT command to send, the AT command wait for, and
+  a timeout.
+
+  The default timeout is 500 milliseconds
+  """
+  @spec send(binary(), binary(), binary(), non_neg_integer()) :: {:ok, [binary()]}
+  def send(tty, command, wait_for, timeout \\ 500) do
+    GenServer.call(server_name(tty), {:send, command, wait_for, timeout})
+  end
+
+  @impl true
+  def init(opts) do
+    speed = Keyword.get(opts, :speed, 115_200)
+    tty = Keyword.fetch!(opts, :tty)
+
+    {:ok, uart} = UART.start_link()
+
+    :ok =
+      UART.open(uart, tty,
+        speed: speed,
+        framing: {UART.Framing.Line, separator: "\r\n"}
+      )
+
+    {:ok,
+     %State{
+       uart: uart,
+       tty: tty
+     }}
+  end
+
+  @impl true
+  def handle_call({:send, command, wait_for, timeout}, from, %{waiter: nil} = state) do
+    timeout_ref = write_at_command(state.uart, command, timeout)
+
+    {:noreply, %{state | waiter: from, stop_response: wait_for, command_timeout_ref: timeout_ref}}
+  end
+
+  def handle_call({:send, command, wait_for, timeout}, from, state) do
+    new_queue = :queue.in({command, wait_for, from, timeout}, state.queue)
+
+    {:noreply, %{state | queue: new_queue}}
+  end
+
+  @impl true
+  def handle_info({:circuits_uart, _tty, ""}, state), do: {:noreply, state}
+
+  def handle_info({:circuits_uart, _tty, at_response}, state) do
+    if state.stop_response == at_response do
+      _ = Process.cancel_timer(state.command_timeout_ref)
+      GenServer.reply(state.waiter, {:ok, state.buffer})
```

Comment:
Does the buffer need to be reversed?

Reply:
I thought about that too and I wasn't sure.

It is really convient to have the last message be the head of the list, so that is why I didn't reverse it. But I can see the case for reversing.